sk-csharp-chatgpt-plugin/semantic-functions-generator/SemanticSkillGenerator.cs

```
@@ -0,0 +1,177 @@
+        var functionFiles = context.AdditionalFiles.Where(f =>
+            f.Path.Contains(FunctionConfigFilename) ||
+            f.Path.Contains(FunctionPromptFilename));
+
+        // Group first by function name, then by parent folder
+        var fnFileGroup = functionFiles.GroupBy(f => Path.GetDirectoryName(f.Path));
+
+        // Group the files by parent folder name
+        var folderGroups = fnFileGroup.GroupBy(f => Path.GetFileName(Path.GetDirectoryName(f.Key)));
+
+        // Generate a class for each folder
+        foreach (var folderGroup in folderGroups)
+        {
+            string? folderName = folderGroup.Key;
+            if (string.IsNullOrWhiteSpace(folderName))
+            {
+                continue;
+            }
+
+            string classSource = GenerateClassSource(rootNamespace!, folderName, folderGroup);
+            context.AddSource(folderName, SourceText.From(classSource, Encoding.UTF8));
+        }
+    }
+
+    // Generate the source code for a folder of semantic functions
+    private static string GenerateClassSource(string rootNamespace, string folderName, IGrouping<string, IGrouping<string, AdditionalText>> folderGroup)
+    {
+        // Use a StringBuilder to build the class source
+        StringBuilder functionsCode = new();
+
+        foreach (var functionGroup in folderGroup)
+        {
+            // Get the "skprompt.txt" and "config.json" files for this function
+            AdditionalText? configFile = functionGroup.FirstOrDefault(f => Path.GetFileName(f.Path).Equals(FunctionConfigFilename, StringComparison.InvariantCultureIgnoreCase));
+            AdditionalText? promptFile = functionGroup.FirstOrDefault(f => Path.GetFileName(f.Path).Equals(FunctionPromptFilename, StringComparison.InvariantCultureIgnoreCase));
+            if (promptFile != default && configFile != default)
+            {
+                functionsCode.AppendLine(GenerateFunctionSource(promptFile, configFile) ?? string.Empty);
```

If we get a `null` response we still proceed as if everything is working. We should probably throw an exception and fail in a way the user knows what is gone wrong and knows what to fix.